torch/csrc/jit/passes/device_type_analysis.cpp

```diff
@@ -0,0 +1,237 @@
+#include <ATen/core/interned_strings.h>
+#include <ATen/core/jit_type.h>
+#include <c10/core/Device.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/device_type_analysis.h>
+#include <torch/library.h>
+#include <memory>
+#include "ATen/core/List.h"
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+using Tensor = at::Tensor;
+using Device = at::Device;
+
+using PropRule = std::function<bool(Node*)>;
+/*
+A Propagation Rule takes the Node, and
+applies the relevant properties to the Tensor outputs
+of the Node (based on the rule itself)
+
+Returns: Bool indicating if anything was changed
+*/
+
+bool setDeviceType(Value* value, Device device, bool can_overwrite = false) {
+  auto tensor_type = value->type()->cast<TensorType>();
+  TORCH_INTERNAL_ASSERT(tensor_type, "Expecting a tensor type");
+  if (!tensor_type->device().has_value()) {
+    value->setType(tensor_type->withDevice(device));
+    return true;
+  }
+  if (tensor_type->device().value() != device) {
+    TORCH_INTERNAL_ASSERT(
+        can_overwrite,
+        "Expected device to be ",
+        device,
+        " but found ",
+        tensor_type->device().value());
+    value->setType(tensor_type->withDevice(device));
+    return true;
+  }
+  return false;
+}
+
+bool setReturnsToDevice(Node* n, Device device) {
+  bool changed = false;
+  for (Value* out : n->outputs()) {
+    auto tensor_type = out->type()->cast<TensorType>();
+    if (!tensor_type) {
+      continue;
+    }
+    changed |= setDeviceType(out, device);
+  }
+  return changed;
+}
+
+bool propWithNoDevice(Node* n) {
+  // Figure out what the common device to propagate is
+  c10::optional<Device> device;
+  for (Value* inp : n->inputs()) {
+    auto tensor_type = inp->type()->cast<TensorType>();
+    if (!tensor_type) {
+      continue;
+    }
+    if (device) {
+      if (tensor_type->device().has_value()) {
+        auto cur_device = tensor_type->device().value();
+        TORCH_INTERNAL_ASSERT(
+            device == cur_device,
+            "Expected device to be ",
+            device.value(),
+            " but found ",
+            cur_device);
+      }
+      continue;
+    } else {
+      device = tensor_type->device();
+    }
+  }
+
+  return device.has_value() && setReturnsToDevice(n, device.value());
+}
+
+
+bool isDeviceArgumentType(const TypePtr type) {
+  // Have to deal with Union Types (and Optional Types are treated as a Union type)
+  for (auto& subtype: type->containedTypes()){
+    if (isDeviceArgumentType(subtype)){
+      return true;
+    }
+  }
+  return DeviceObjType::get() == type;
+}
+
+bool defaultDeviceProp(Node* n) {
+  // Detecting of the op has a device object argument
+  // as there is implicit string conversion to device
+  auto arguments = n->getOperator().schema().arguments();
+  auto input_vec = n->inputs();
```

Comment:
Do we need to handle the zero-dim case here?  `torch.rand([4, 4]).cuda() + torch.tensor(0.1)` ? The device is not taken from inputs

Reply:
This would be part of the `propWithNoDevice` function, but ugh, I didn't realize that there was that edge case.

Reply:
I think we can update the heuristics to something along the line of the following
a) take the device of a tensor if it has size set and it's dim > 0
b) if all tensors have the same device, take it

Reply:
I think we also need to exclude list :
reshape_as,
view_as,
expand_as

Reply:
Currently they should be properly handled by the logic of "return no device if there are multiple devices in the input tensors" logic. However yes, we should add more support for them later.